strapi/strapi
# feature/be-telemetry

### What does it do?

This PR introduces a new telemetry package to Strapi. This package is intended to be a single source of truth for our telemetry and analytics needs. This PR is the first one in a series of 3, as the whole process of introducing the new telemetry package has been broken up into 3 parts. 

One of the new things that has been brought in to Strapi telemetry is the use of Amplitude Data (formerly known as Iteratively). It allows us to form a defined tracking plan and validate events against it. As such, all events that have been tracked by `strapi.telemetry.send()` have been updated to use ampli instead in this PR.

What has been done:

1. Original `metrics` code located in `/packages/core/strapi` has been removed and migrated to the new `@strapi/telemetry-be` package. Logic remained the same: when an instance of Strapi object is created, and instance of the telemetry is created at the same time, with sender and rate limiter and cron lifecycle preserved. The only difference is that the code comes from a different place now

2. All events that were tracked using `strapi.telemetry.send()/this.telemetry.send()` now use the following format: `ampli.[nameOfEvent]()`. As such, `ampli` is now imported everywhere, where such event tracking occurs. 

3. An instance of Ampli is loaded when telemetry package is loaded first time. An instance of Ampli also has 1 middleware attached to it. This middleware allows us to prevent sending tracking data directly to Amplitude and instead rerouting it to our tracker. It is important to note that changing the current logic of sending tracked events was not in the scope of the initiative, so the way ampli validation works is as follows:
  - `ampli.[nameOfEvent]()` accepts 4 arguments, last of which is an argument called `extra`. This info is passed directly to any middleware that are attached to ampli.
  - therefore, each ampli event tracking occurrence is expecting an object to know where the event originates from and a the function to use in order to send the event onward. Example: `ampli.didStartServer('', {}, {}, { source: 'core', send: strapi.telemetry.send})`
  - Middleware then will use this `send` function in order to route the event through the metrics logic as before. The reason it is done this way is because for this iteration of the package it will allow us to standardize the way we call this new tracking and test the whole implementation without having to change much logic. We have 3 different ways we track events, and as such this approach is the least destructive one and less prone to error.

### Why is it needed?

We were planning to introduce Amplitude Data (formerly Iteratively) for some time now, but were delaying it due to changes within Amplitude itself. Since the product is stable now, we can finally start implementing it. This will allow us to manage out tracking plan easier and also decrease the amount of time needed for a developer to introduce new tracked events.

### How to test it?

You can test this branch by running the getstarted project. All events should be logged out to the console. Also, the sender url has been changed to `http://localhost:3300` for now, so you can start a simple express/koa server with a `/track` POST endpoint. You should see events received there as well (note that in this case events will be reformatted to fit tracker requirements).

This is a draft PR for now, all console logs will be removed and urls fixed once this PR is looked at by other team members.
